src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -234,15 +249,340 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+        extensions: &Extensions<'_>,
+    ) -> Result<(), ValidationError> {
+        self.permits_basic(cert)?;
+
+        for ext_policy in self.ee_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether `issuer` is a valid issuing CA for `child` at a
+    /// path-building depth of `current_depth`.
+    ///
+    /// This checks that `issuer` is permitted under this policy and that
+    /// it was used to sign for `child`.
+    ///
+    /// On success, this function returns the new path-building depth. This
+    /// may or may not be a higher number than the original depth, depending
+    /// on the kind of validation performed (e.g., whether the issuer was
+    /// self-issued).
+    pub(crate) fn valid_issuer(
+        &self,
+        issuer: &Certificate<'_>,
+        child: &Certificate<'_>,
+        current_depth: u8,
+        issuer_extensions: &Extensions<'_>,
+    ) -> Result<u8, ValidationError> {
+        // The issuer needs to be a valid CA at the current depth.
+        self.permits_ca(issuer, current_depth, issuer_extensions)?;
+
+        // CA/B 7.1.3.1 SubjectPublicKeyInfo
+        if !self
+            .permitted_public_key_algorithms
+            .contains(&child.tbs_cert.spki.algorithm)
+        {
+            return Err(ValidationError::Other(format!(
+                "Forbidden public key algorithm: {:?}",
+                &child.tbs_cert.spki.algorithm
+            )));
+        }
+
+        // CA/B 7.1.3.2 Signature AlgorithmIdentifier
+        if !self
+            .permitted_signature_algorithms
+            .contains(&child.signature_alg)
+        {
+            return Err(ValidationError::Other(format!(
+                "Forbidden signature algorithm: {:?}",
+                &child.signature_alg
+            )));
+        }
+
+        let pk = self
+            .ops
+            .public_key(issuer)
+            .map_err(|_| ValidationError::Other("issuer has malformed public key".to_string()))?;
+        if self.ops.verify_signed_by(child, pk).is_err() {
+            return Err(ValidationError::Other(
+                "signature does not match".to_string(),
+            ));
+        }
+
+        // Self-issued issuers don't increase the working depth.
+        match cert_is_self_issued(issuer) {
+            true => Ok(current_depth),
+            false => Ok(current_depth.checked_add(1).ok_or_else(|| {
+                ValidationError::Other("current depth calculation overflowed".to_string())
+            })?),
         }
     }
 }
 
+fn permits_validity_date(validity_date: &Time) -> Result<(), ValidationError> {
+    const GENERALIZED_DATE_CUTOFF_YEAR: u16 = 2050;
+
+    // NOTE: The inverse check on `asn1::UtcTime` is already done for us
+    // by the variant's constructor.
+    if let Time::GeneralizedTime(_) = validity_date {
+        if validity_date.as_datetime().year() < GENERALIZED_DATE_CUTOFF_YEAR {
```

Comment:
This couldn't matter less, but this is technically wrong for a cert with issuance date in 1949 or older.

Reply:
This should be fixed.